README.md
# SRAGPU-nf
This is a Nextflow workflow that demonstrates the use of GPU enabled Singularity containers. The two sample processing workflows `main.nf` and `nvBowtie_WF.nf` utilize GPU enable short read aligners, BarraCUDA and nvBowtie respectively. Additionally, `classify.nf` is a single step machine learning classification to illustrate that TensorFlow Docker Hub images can be pulled and automatically converted to Singularity images. The GPU pass through will still work for containers generated from these images.

Additional information is available at: https://github.com/KevinSayers/MastersThesis

# Requirements
* Nextflow version >= 0.25.5
* Singularity

# GPU pass through
Singularity will automatically mount the necessary NVIDIA CUDA libraries into containers if the `--nv` flag is specified. This can be configured in the `nextflow.config` as shown below.

```
singularity{
	enabled = true
	runOptions = '--nv'
}
```

# Short read aligners
Two workflows are provided to demonstrate the use of Singularity containers with GPU short read aligners.

To run either first run `nextflow run setup.nf` this will download and extract the necessary reference sequence. The first alinger is BarraCUDA, this aligner can be used by executing `nextflow run main.nf` . The other aligner is the nvBowtie, a version of Bowtie2 developed by NVIDIA. This aligner can be run using `nextflow run nvBowtie_WF.nf`. Either workflow will automatically pull the relevant Singularity container from a publicly available repository. 

# TensorFlow
The `classify.nf` workflow also shows the pulling of a GPU enabled Docker Hub image. This workflow will automatically pull the latest Python 3 TensorFlow GPU Docker image. It then trains a SVM model to classify samples based on one-hot encoded variants in the `transcriptomepreprocessed` file. Nextflow automatically converts the Docker Hub image to Singularity, and the GPU pass through enables the container to be used. 




.nextflow/history
2017-09-27 21:04:03	3m 24s	cocky_spence	OK	dbb0011f60f4a43e39d4b20880c4a58b	3302c465-ad72-4d98-8553-11d4b4219c12	nextflow run classify.nf
2017-09-27 21:09:46	1m 6s	ridiculous_bell	ERR	960ddae40419fa99352c585a2bfde4cf	b5fe5a52-e563-4cae-83b8-62c67f5ad2a3	nextflow run main.nf


nn.py
#!python3
import tensorflow as tf 
import numpy as np 
import random
import sys
import matplotlib.pyplot as plt


def runModel(encodedFile):
	data = open(encodedFile,'r')
	temp = []

	for i in data:
		temp.append(i.strip().split(','))
	#random.shuffle(temp)

	trainingdata = []
	traininglabels = []
	counttrain = 0
	# BELOW HERE IS DIFFERENT
	for j in temp[0:342]:
		trainingdata.append(np.asarray(j[2:]))
		if j[1] == "triple-negative breast cancer (TNBC)":
			traininglabels.append(np.asarray([1,0]))
			counttrain += 1
		else:
			traininglabels.append(np.asarray([0,1]))

	testdata = []
	testlabels = []
	counttest= 0
	for j in temp[343:]:
		testdata.append(np.asarray(j[2:]))
		if j[1] == "triple-negative breast cancer (TNBC)":
			testlabels.append(np.asarray([1,0]))
			counttest += 1
		else:
			testlabels.append(np.asarray([0,1]))
	print(counttrain)
	print(counttest)


	print (len(trainingdata[0]))
	inputs = len(trainingdata[0])
	classes = 2
	hidden = 250

	traininglabels = np.asarray(traininglabels)
	traininglabels = np.reshape(traininglabels,(len(traininglabels),classes))

	testlabels = np.asarray(testlabels)
	testlabels = np.reshape(testlabels,(len(testlabels),classes))

	trainingdata = np.asarray(trainingdata)
	testdata = np.asarray(testdata)
	testdata = testdata.astype(int)
	trainingdata = trainingdata.astype(int)


	x = tf.placeholder(tf.float32, [None, len(trainingdata[0])])
# W = tf.Variable(tf.zeros([540728, 2]))
	W = tf.Variable(tf.truncated_normal([inputs, 2], stddev=0.1))
	b = tf.Variable(tf.constant(0.5))

	y = tf.matmul(x, W) + b
	y_ = tf.placeholder(tf.float32, [None, 2])

	cross_entropy = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits(labels=y_, logits=y))
	train_step = tf.train.GradientDescentOptimizer(0.0005).minimize(cross_entropy)

	correct_prediction = tf.equal(tf.argmax(y, 1), tf.argmax(y_, 1))
	accuracy = tf.reduce_mean(tf.cast(correct_prediction, tf.float32))

	
	alltrain = []
	alltst = []
	allloss = []

	for j in range(0,5):
		sess = tf.InteractiveSession()
		tf.global_variables_initializer().run()
		datapoints = []
		testing = []
		loss = []

		for _ in range(20):
			#print (tf.Tensor.eval(wout))
			sess.run(train_step, feed_dict={x: trainingdata, y_: traininglabels})
			
			datapoints.append(sess.run(accuracy, feed_dict={x: trainingdata, y_: traininglabels}))
			testing.append(sess.run(accuracy, feed_dict={x: testdata, y_: testlabels}))
			loss.append(sess.run(cross_entropy, feed_dict={x: testdata, y_: testlabels}))

		finallabels = sess.run(y, feed_dict={x: testdata})
		tnbc_true = 0
		tnbc_false = 0
		tnbc_true_neg = 0 
		tnbc_false_neg = 0
		
		for i in range(0, len(testlabels)):
			actual = testlabels[i].argmax(axis=0)
			testresult = finallabels[i].argmax(axis=0)
			if actual == 0 and testresult == 0:
				tnbc_true += 1
			if actual == 0 and testresult == 1:
				tnbc_false += 1
			if actual == 1 and testresult == 1:
				tnbc_true_neg += 1
			if actual == 1 and testresult == 0:
				tnbc_false_neg += 1
		sensitivity = tnbc_true/(tnbc_true+tnbc_false)
		specificity = tnbc_true_neg/(tnbc_true_neg+tnbc_false_neg)
		print (tnbc_true)
		print (tnbc_false)
		print ("Sensitivity:%s"%(sensitivity))
		print ("Specificity:%s"%(specificity))

		alltrain.append(datapoints)
		alltst.append(testing)
		allloss.append(loss)
	
	plt.title('Training accuracy')
	plt.ylabel('Accuracy')
	plt.xlabel('iteration')
	for i in alltrain:
		plt.plot(i)
	plt.show()

	plt.title('Training loss')
	plt.ylabel('loss')
	plt.xlabel('iteration')
	for i in allloss:
		plt.plot(i)
	plt.show()

	
	plt.title('Testing accuracy')
	plt.ylabel('Accuracy')
	plt.xlabel('iteration')
	for i in alltst:
		plt.plot(i)
	plt.show()
	
	# plt.plot(datapoints)
	# plt.show()
	# plt.title('Training loss')
	# plt.ylabel('loss')
	# plt.xlabel('iteration')
	# plt.plot(loss)
	# plt.show()
	# plt.title('Testing accuracy')
	# plt.ylabel('Accuracy')
	# plt.xlabel('iteration')
	# plt.plot(testing)
	# plt.show()

	# counter = 1
	# for i in testlabels:
	# 	print (str(counter) + str(i))
	# 	counter += 1

def main():
	#runModel(sys.argv[1])
	runModel('transcriptomepreprocessed')

if __name__ == "__main__":
	main()

svm.py
# import matplotlib.pyplot as plt
import numpy as np
import tensorflow as tf
import random
import pickle
# Create graph
sess = tf.Session()



# Declare batch size
batch_size = 100

data = open('transcriptomepreprocessed','r')
temp = []

for i in data:
    temp.append(i.strip().split(','))
#random.shuffle(temp)

trainingdata = []
traininglabels = []

# BELOW HERE IS DIFFERENT
for j in temp[0:342]:
    trainingdata.append(np.asarray(j[2:]))
    if j[1] == "triple-negative breast cancer (TNBC)":
        traininglabels.append(np.asarray([1]))
    else:
        traininglabels.append(np.asarray([-1]))

testdata = []
testlabels = []
for j in temp[343:]:
    testdata.append(np.asarray(j[2:]))
    if j[1] == "triple-negative breast cancer (TNBC)":
        testlabels.append(np.asarray([1]))
    else:
        testlabels.append(np.asarray([-1]))


print (len(trainingdata[0]))
inputs = len(trainingdata[0]) #12360
classes = 2
hidden = 500

traininglabels = np.asarray(traininglabels)
traininglabels = np.reshape(traininglabels,(len(traininglabels),1))

testlabels = np.asarray(testlabels)
testlabels = np.reshape(testlabels,(len(testlabels),1))

trainingdata = np.asarray(trainingdata)
testdata = np.asarray(testdata)
testdata = testdata.astype(int)
trainingdata = trainingdata.astype(int)

# Initialize placeholders
x = tf.placeholder(tf.float32, [None, inputs])
y_ = tf.placeholder(tf.float32, [None, 1])

# Create variables for linear regression
A = tf.Variable(tf.random_normal(shape=[inputs,1]))
b = tf.Variable(tf.random_normal(shape=[1,1]))

# Declare model operations
model_output = tf.subtract(tf.matmul(x, A), b)

# Declare vector L2 'norm' function squared
l2_norm = tf.reduce_sum(tf.square(A))

# Declare loss function
# Loss = max(0, 1-pred*actual) + alpha * L2_norm(A)^2
# L2 regularization parameter, alpha
alpha = tf.constant([0.001])
# Margin term in loss
classification_term = tf.reduce_mean(tf.maximum(0., tf.subtract(1., tf.multiply(model_output, y_))))
# Put terms together
loss = tf.add(classification_term, tf.multiply(alpha, l2_norm))

# Declare prediction function
prediction = tf.sign(model_output)
accuracy = tf.reduce_mean(tf.cast(tf.equal(prediction, y_), tf.float32))

# Declare optimizer
my_opt = tf.train.AdamOptimizer(0.0005)
train_step = my_opt.minimize(loss)

# Initialize variables


# Training loop

alltrain = []
alltst = []
allloss = []
for j in range(0,5):
    init = tf.global_variables_initializer()
    sess.run(init)

    loss_vec = []
    train_accuracy = []
    test_accuracy = []
    losspts = []

    for i in range(2000):
        sess.run(train_step, feed_dict={x: trainingdata, y_: traininglabels})

        temp_loss = sess.run(loss, feed_dict={x: trainingdata, y_: traininglabels})
        loss_vec.append(temp_loss)

        train_acc_temp = sess.run(accuracy, feed_dict={
            x: trainingdata,
            y_: traininglabels})
        train_accuracy.append(train_acc_temp)

        test_acc_temp = sess.run(accuracy, feed_dict={
            x: testdata,
            y_: testlabels})
        test_accuracy.append(test_acc_temp)

    alltrain.append(train_accuracy)
    allloss.append(loss_vec)
    alltst.append(test_accuracy)

# plt.title('Training accuracy')
# plt.ylabel('Accuracy')
# plt.xlabel('iteration')
# for i in alltrain:
#     plt.plot(i)
# plt.savefig('finalsvmtrain.png')
# plt.clf()

# plt.title('Training loss')
# plt.ylabel('loss')
# plt.xlabel('iteration')
# for i in allloss:
#     plt.plot(i)
# plt.savefig('finalsvmloss.png')
# plt.clf()


# plt.title('Testing accuracy')
# plt.ylabel('Accuracy')
# plt.xlabel('iteration')
# for i in alltst:
#     plt.plot(i)
# plt.savefig('finalsvmtest.png')
# plt.clf()

nnmutli.py
#!python3
import tensorflow as tf 
import numpy as np 
import random
import sys
import matplotlib.pyplot as plt


def runModel(encodedFile):
	data = open(encodedFile,'r')
	temp = []

	for i in data:
		temp.append(i.strip().split(','))
	#random.shuffle(temp)

	trainingdata = []
	traininglabels = []
	counttrain = 0
	# BELOW HERE IS DIFFERENT
	for j in temp[0:342]:
		trainingdata.append(np.asarray(j[2:]))
		if j[1] == "triple-negative breast cancer (TNBC)":
			counttrain += 1
			traininglabels.append(np.asarray([1,0,0,0]))
		if j[1] == "human epidermal growth factor receptor 2 positive (HER2+)":
			traininglabels.append(np.asarray([0,1,0,0]))
		if j[1] == "estrogen receptor positive (ER+)":
			traininglabels.append(np.asarray([0,0,1,0]))
		if j[1] == "double positive (ER+ and HER2+)":
			traininglabels.append(np.asarray([0,0,0,1]))

	testdata = []
	testlabels = []
	counttest= 0
	for j in temp[343:]:
		testdata.append(np.asarray(j[2:]))
		if j[1] == "triple-negative breast cancer (TNBC)":
			counttest += 1
			testlabels.append(np.asarray([1,0,0,0]))
		if j[1] == "human epidermal growth factor receptor 2 positive (HER2+)":
			testlabels.append(np.asarray([0,1,0,0]))
		if j[1] == "estrogen receptor positive (ER+)":
			testlabels.append(np.asarray([0,0,1,0]))
		if j[1] == "double positive (ER+ and HER2+)":
			testlabels.append(np.asarray([0,0,0,1]))

	print(counttrain)
	print(counttest)


	print (len(trainingdata[0]))
	print (len(testlabels))
	print (len(testdata))
	inputs = len(trainingdata[0])
	classes = 4
	hidden = 250

	traininglabels = np.asarray(traininglabels)
	traininglabels = np.reshape(traininglabels,(len(traininglabels),classes))

	testlabels = np.asarray(testlabels)
	testlabels = np.reshape(testlabels,(len(testlabels),classes))

	trainingdata = np.asarray(trainingdata)
	testdata = np.asarray(testdata)
	testdata = testdata.astype(int)
	trainingdata = trainingdata.astype(int)


	x = tf.placeholder(tf.float32, [None, len(trainingdata[0])])
# W = tf.Variable(tf.zeros([540728, 2]))
	W = tf.Variable(tf.truncated_normal([inputs, classes], stddev=0.1))
	b = tf.Variable(tf.constant(0.5))

	y = tf.matmul(x, W) + b
	y_ = tf.placeholder(tf.float32, [None, classes])

	cross_entropy = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits(labels=y_, logits=y))
	train_step = tf.train.GradientDescentOptimizer(0.0005).minimize(cross_entropy)

	correct_prediction = tf.equal(tf.argmax(y, 1), tf.argmax(y_, 1))
	accuracy = tf.reduce_mean(tf.cast(correct_prediction, tf.float32))

	
	alltrain = []
	alltst = []
	allloss = []

	for j in range(0,5):
		sess = tf.InteractiveSession()
		tf.global_variables_initializer().run()
		datapoints = []
		testing = []
		loss = []

		for _ in range(20000):
			#print (tf.Tensor.eval(wout))
			sess.run(train_step, feed_dict={x: trainingdata, y_: traininglabels})
			
			datapoints.append(sess.run(accuracy, feed_dict={x: trainingdata, y_: traininglabels}))
			testing.append(sess.run(accuracy, feed_dict={x: testdata, y_: testlabels}))
			loss.append(sess.run(cross_entropy, feed_dict={x: testdata, y_: testlabels}))

		# finallabels = sess.run(y, feed_dict={x: testdata})
		# tnbc_true = 0
		# tnbc_false = 0
		# tnbc_true_neg = 0 
		# tnbc_false_neg = 0
		
		# for i in range(0, len(testlabels)):
		# 	actual = testlabels[i].argmax(axis=0)
		# 	testresult = finallabels[i].argmax(axis=0)
		# 	if actual == 0 and testresult == 0:
		# 		tnbc_true += 1
		# 	if actual == 0 and testresult == 1:
		# 		tnbc_false += 1
		# 	if actual == 1 and testresult == 1:
		# 		tnbc_true_neg += 1
		# 	if actual == 1 and testresult == 0:
		# 		tnbc_false_neg += 1
		# sensitivity = tnbc_true/(tnbc_true+tnbc_false)
		# specificity = tnbc_true_neg/(tnbc_true_neg+tnbc_false_neg)
		# print (tnbc_true)
		# print (tnbc_false)
		# print ("Sensitivity:%s"%(sensitivity))
		# print ("Specificity:%s"%(specificity))

		alltrain.append(datapoints)
		alltst.append(testing)
		allloss.append(loss)
	
	plt.title('Training accuracy')
	plt.ylabel('Accuracy')
	plt.xlabel('iteration')
	for i in alltrain:
		plt.plot(i)
	plt.show()

	plt.title('Training loss')
	plt.ylabel('loss')
	plt.xlabel('iteration')
	for i in allloss:
		plt.plot(i)
	plt.show()

	
	plt.title('Testing accuracy')
	plt.ylabel('Accuracy')
	plt.xlabel('iteration')
	for i in alltst:
		plt.plot(i)
	plt.show()
	
	# plt.plot(datapoints)
	# plt.show()
	# plt.title('Training loss')
	# plt.ylabel('loss')
	# plt.xlabel('iteration')
	# plt.plot(loss)
	# plt.show()
	# plt.title('Testing accuracy')
	# plt.ylabel('Accuracy')
	# plt.xlabel('iteration')
	# plt.plot(testing)
	# plt.show()

	# counter = 1
	# for i in testlabels:
	# 	print (str(counter) + str(i))
	# 	counter += 1

def main():
	#runModel(sys.argv[1])
	runModel('transcriptomepreprocessed')

if __name__ == "__main__":
	main()